### DISCUSSION TITLE
Does Quartz 3 support Read Committed Snapshot Isolation when using Sql Server (2019)

### DISCUSSION THREAD

**Discussion by IDDesigns** [quartznet/quartznet#2622] [General]
We heavily use Quartz and are currently finding that (on SQL Server) it is currently the highest usage in terms of CPU and logical reads - it has been suggested we turn on RSCI for the Quartz database, is there any reason that we shouldn't do this?

Many thanks

Ian

**Comment by lahma** [comment]
I wouldn't jump into that without checking what Quartz version I'm using and why is the database so under stress.

**Reply by IDDesigns** [reply]
It's the latest Quartz version (3.13.1). There is a single quartz database which is used by around 200 websites running the same software (unique scheduler name per site), each of those running as clustered with 3 instances. Each site has around 20-30 jobs running from 3 second intervals up to once every 24 hours. About 37 million writes to Qrtz_triggers per day.

**Reply by lahma** [reply]
And you are using the default MSSQLServerDelegate configuration etc locking, like just the DI configuration syntax without customization?

**Reply by lahma** [reply]
The thing here is that you have very little context to your question. But generally I would advice against snapshot isolation as Quartz uses database-based locks for coordinate in cluster.

Because without context, an easy win would run your jobs every 6 seconds instead of 3 seconds, you just doubled your capacity.

**Reply by IDDesigns** [reply]
Apologies for the lack to context, I had thought the question about RCSI would be black and white.

Our core configuration is as follows:

`        services.Configure<QuartzOptions>(
            options =>
            {
                options.Scheduling.OverWriteExistingData = true;
                options.Scheduling.IgnoreDuplicates = true;
                options.Scheduling.ScheduleTriggerRelativeToReplacedTrigger = true;
            });

        services.AddQuartz(
            q =>
            {
                q.SchedulerName = instanceName;
                q.SchedulerId = "AUTO";
                q.MisfireThreshold = TimeSpan.FromMinutes(10);
                q.UseDefaultThreadPool(
                    tp =>
                    {
                        tp.MaxConcurrency = 5;
                    });

                q.UsePersistentStore(
                    s =>
                    {
                        s.UseProperties = true;
                        s.RetryInterval = TimeSpan.FromSeconds(15);
                        s.UseSqlServer(
                            sqlServer =>
                            {
                                sqlServer.ConnectionString = connectionString;
                                sqlServer.TablePrefix = @"QRTZ_";
                            });
                        s.UseNewtonsoftJsonSerializer();
                        s.UseClustering(c =>
                        {
                            c.CheckinMisfireThreshold = TimeSpan.FromMinutes(5);
                            c.CheckinInterval = TimeSpan.FromMinutes(5);
                        });
                    });

                q.InterruptJobsOnShutdown = true;
                q.InterruptJobsOnShutdownWithWait = true;
            });`

And the .net framework side:

`        var props = new NameValueCollection
        {
            { "quartz.scheduler.instanceName", instanceName,
            { "quartz.scheduler.instanceId", "AUTO" },
            { "quartz.threadPool.threadCount", "5" }, // 5 job can run at the same time
            { "quartz.jobStore.misfireThreshold", "60000" },
            { "quartz.jobStore.type", "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz" },
            { "quartz.jobStore.clustered", "true" },
            { "quartz.serializer.type", "json" },
            { "quartz.jobStore.driverDelegateType", "Quartz.Impl.AdoJobStore.SqlServerDelegate, Quartz" },
            { "quartz.jobStore.tablePrefix", @"QRTZ_" },
            { "quartz.jobStore.useProperties", "true" },
            { "quartz.jobStore.dataSource", "lms" },
            { "quartz.dataSource.lms.connectionString", connectionString },
            { "quartz.dataSource.lms.provider", "SqlServer" }
        };`

**Reply by lahma** [reply]
Configuration looks OK to me (max concurrency is quite low though). So to add some more context, is your database dying due to the load, is it some DTU-limited cloud resource, can you give it more resources? Or is this just that it's using 20% of resources and you would want it to take 10%?

**Reply by IDDesigns** [reply]
Quartz is using 40% of the Sql Server CPU and accounts for 70% of the total logical reads on the server so we a looking to reduce it's load on the server, ideally without compromising the number of jobs we are running. 
Master accounts for another 40% of cpu and the other 120 databases vary a bit but typically vary between 0.01 and 5% in bursts. In terms of logical reads all the other databases are < 5% of the total (again with some variation from 0.01 to 5%)

**Reply by lahma** [reply]
You could experiment with `quartz.scheduler.batchTriggerAcquisitionMaxCount` (default 1) to get more triggers to handle (max concurrency will limit too, so lowest chosen). This will make Quartz to take more triggers on each pass which can change the system load behavior. Quartz might be taking a lot of resources but it seems that it's also doing a lot of work based on your numbers. Maybe your other logic just isn't that often run?

**Reply by IDDesigns** [reply]
Thanks I'll try that. 

The load on the other databases is driven from the functionality of the Quartz jobs and from web requests/api calls, we currently have somewhere around 90K requests per minute across the estate so there are plenty of other db requests going on (and we are looking at the load as well)

**Reply by IDDesigns** [reply]
We have made a few tweaks to the Quartz database - dropped all the indexes, added statistics and tweak the primary key indexes - this has resulted in the CPU for the Quartz database dropping to 0.01% of the CPU with no change in usage - this is on Sql 2019
The drop statements were taken from the sql in the Quartz repo so have some of the legacy indexes in too. It looks like a lot of the indexes were written to a lot but never read so were creating a lot of overhead.

The script we used is:

`DROP INDEX IF EXISTS [IDX_QRTZ_T_J] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_JG] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_C] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_G] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_G_J] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_STATE] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_N_STATE] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_N_G_STATE] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_NEXT_FIRE_TIME] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_NFT_ST] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_NFT_MISFIRE] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_NFT_ST_MISFIRE] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_T_NFT_ST_MISFIRE_GRP] ON [dbo].[QRTZ_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_FT_TRIG_INST_NAME] ON [dbo].[QRTZ_FIRED_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_FT_INST_JOB_REQ_RCVRY] ON [dbo].[QRTZ_FIRED_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_FT_J_G] ON [dbo].[QRTZ_FIRED_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_FT_JG] ON [dbo].[QRTZ_FIRED_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_FT_T_G] ON [dbo].[QRTZ_FIRED_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_FT_TG] ON [dbo].[QRTZ_FIRED_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_FT_G_J] ON [dbo].[QRTZ_FIRED_TRIGGERS];
GO
DROP INDEX IF EXISTS [IDX_QRTZ_FT_G_T] ON [dbo].[QRTZ_FIRED_TRIGGERS];
GO

CREATE STATISTICS SCHED_NAME_MISFIRE_INSTR_NEXT_FIRE_TIME ON [dbo].[QRTZ_TRIGGERS] ( [SCHED_NAME] ,[MISFIRE_INSTR] ,[NEXT_FIRE_TIME],[TRIGGER_STATE] )  WITH SAMPLE 30 PERCENT, PERSIST_SAMPLE_PERCENT = ON;
GO
CREATE STATISTICS SCHED_NAME_JOB_NAME_JOB_GROUP ON [dbo].[QRTZ_TRIGGERS] ([SCHED_NAME] ,	[JOB_NAME] ,	[JOB_GROUP] )  WITH SAMPLE 30 PERCENT, PERSIST_SAMPLE_PERCENT = ON;
GO
CREATE STATISTICS SCHED_NAME_TRIGGER_GROUP ON [dbo].[QRTZ_TRIGGERS] ( [SCHED_NAME] ,[TRIGGER_GROUP] )  WITH SAMPLE 30 PERCENT, PERSIST_SAMPLE_PERCENT = ON;
GO
CREATE STATISTICS SCHED_NAME_JOB_NAME_JOB_GROUP ON [dbo].[QRTZ_FIRED_TRIGGERS] ( [SCHED_NAME] ,	[JOB_NAME] ,	[JOB_GROUP]  )  WITH SAMPLE 30 PERCENT, PERSIST_SAMPLE_PERCENT = ON;
GO
CREATE STATISTICS SCHED_NAME_TRIGGER_GROUP ON [dbo].[QRTZ_FIRED_TRIGGERS] ( [SCHED_NAME] ,	[TRIGGER_GROUP]  )  WITH SAMPLE 30 PERCENT, PERSIST_SAMPLE_PERCENT = ON;
GO
CREATE STATISTICS SCHED_NAME_INSTANCE_NAME_REQUESTS_RECOVERY ON [dbo].[QRTZ_FIRED_TRIGGERS] ( [SCHED_NAME] ,	[INSTANCE_NAME] ,	[REQUESTS_RECOVERY]  )  WITH SAMPLE 30 PERCENT, PERSIST_SAMPLE_PERCENT = ON;
GO

ALTER INDEX [PK_QRTZ_TRIGGERS] ON [dbo].[QRTZ_TRIGGERS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_TRIGGERS] ON [dbo].[QRTZ_TRIGGERS] SET ( OPTIMIZE_FOR_SEQUENTIAL_KEY = ON );
GO
ALTER INDEX [PK_QRTZ_FIRED_TRIGGERS] ON [dbo].[QRTZ_FIRED_TRIGGERS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_FIRED_TRIGGERS] ON [dbo].[QRTZ_FIRED_TRIGGERS] SET ( OPTIMIZE_FOR_SEQUENTIAL_KEY = ON );
GO
ALTER INDEX [PK_QRTZ_SIMPLE_TRIGGERS] ON [dbo].[QRTZ_SIMPLE_TRIGGERS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_SIMPLE_TRIGGERS] ON [dbo].[QRTZ_SIMPLE_TRIGGERS]SET ( OPTIMIZE_FOR_SEQUENTIAL_KEY = ON );
GO
ALTER INDEX [PK_QRTZ_SCHEDULER_STATE] ON [dbo].[QRTZ_SCHEDULER_STATE] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_SCHEDULER_STATE] ON [dbo].[QRTZ_SCHEDULER_STATE]SET ( OPTIMIZE_FOR_SEQUENTIAL_KEY = ON );
GO
ALTER INDEX [PK_QRTZ_CRON_TRIGGERS] ON [dbo].[QRTZ_CRON_TRIGGERS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_CRON_TRIGGERS] ON [dbo].[QRTZ_CRON_TRIGGERS]SET ( OPTIMIZE_FOR_SEQUENTIAL_KEY = ON );
GO

ALTER INDEX [PK_QRTZ_SIMPROP_TRIGGERS] ON [dbo].[QRTZ_SIMPROP_TRIGGERS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_JOB_DETAILS] ON [dbo].[QRTZ_JOB_DETAILS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_LOCKS] ON [dbo].[QRTZ_LOCKS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_PAUSED_TRIGGER_GRPS] ON [dbo].[QRTZ_PAUSED_TRIGGER_GRPS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_BLOB_TRIGGERS] ON [dbo].[QRTZ_BLOB_TRIGGERS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO
ALTER INDEX [PK_QRTZ_CALENDARS] ON [dbo].[QRTZ_CALENDARS] REBUILD PARTITION = ALL WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80);
GO`

**Reply by lahma** [reply]
What version of index scripts were you using, we optimized them some four years ago and seems that you are dropping indexes that are no longer relevant.

https://github.com/quartznet/quartznet/blob/main/database/tables/tables_sqlServer.sql

**Reply by IDDesigns** [reply]
The drop indexes were copied from the quartz sql file you mentioned so they drop both new and old indexes.

**Reply by lahma** [reply]
So did you have the legacy indexes in place?

**Reply by IDDesigns** [reply]
Not on the database we were looking at but we did spot some of our other instances were still the old version.

**Reply by lahma** [reply]
Great, thank you for elaborating. I'm just trying to narrow down how much tweaking existing vs dropping legacy had in play here.

**Reply by lahma** [reply]
Your investigation has been very valuable. I think it would be really helpful if you could pin down what was/were the actual changes that made the difference, like index schema compare.
